## Step 1: Create a SparkSession with a SparkContext

In [1]:
import pyspark
spark = pyspark.sql.SparkSession.builder.getOrCreate()
sc = spark.sparkContext

In [2]:
spark

In [3]:
sc

<SparkContext master=local[*] appName=pyspark-shell>

## Step 2: Create a Spark DataFrame

In [4]:
import pandas as pd

### Ratings Dataframe

In [5]:
#Notice this is a local file. There are computer clusters don't have local files so be careful.
rf = spark.read.json('data/ratings.json')

In [6]:
#look at schema
rf

DataFrame[movie_id: bigint, rating: bigint, timestamp: double, user_id: bigint]

In [7]:
rf.limit(5).show()

+--------+------+------------+-------+
|movie_id|rating|   timestamp|user_id|
+--------+------+------------+-------+
|     858|     4|9.56678732E8|   6040|
|    2384|     4|9.56678754E8|   6040|
|     593|     5|9.56678754E8|   6040|
|    1961|     4|9.56678777E8|   6040|
|    1419|     3|9.56678856E8|   6040|
+--------+------+------------+-------+



In [8]:
rf.limit(5).toPandas()

movie_id  rating    timestamp  user_id
0       858       4  956678732.0     6040
1      2384       4  956678754.0     6040
2       593       5  956678754.0     6040
3      1961       4  956678777.0     6040
4      1419       3  956678856.0     6040

In [9]:
#Instead of loading this file in over again every time we want to look at it, we'll use
rf.persist()

DataFrame[movie_id: bigint, rating: bigint, timestamp: double, user_id: bigint]

In [10]:
rf.head()

Row(movie_id=858, rating=4, timestamp=956678732.0, user_id=6040)

In [11]:
drf = rf.toPandas()

In [12]:
drf.head()

movie_id  rating    timestamp  user_id
0       858       4  956678732.0     6040
1      2384       4  956678754.0     6040
2       593       5  956678754.0     6040
3      1961       4  956678777.0     6040
4      1419       3  956678856.0     6040

In [46]:
drf.loc[drf.user_id == 2060]

movie_id  rating    timestamp  user_id
525945      1721       3  974630327.0     2060
525978      1961       4  974630385.0     2060
526003       260       5  974630451.0     2060
526046      3101       4  974630545.0     2060
526142      3421       4  974630724.0     2060
526174      1270       5  974630766.0     2060
526176      1265       5  974630766.0     2060
526199      1079       3  974630818.0     2060
526229      2716       5  974630858.0     2060
526230      3361       4  974630858.0     2060
526245      2997       3  974630894.0     2060
526254      3671       5  974630923.0     2060
526360       589       5  974631163.0     2060
526361      2028       4  974631163.0     2060
526362      2571       4  974631163.0     2060
526363       457       3  974631163.0     2060
526384      2916       5  974631199.0     2060
526385        21       4  974631199.0     2060
526395       474       2  974631239.0     2060
526396      3418       3  974631239.0     2060
526397      1580       4  974631239.0     2060
526398       377       1  974631239.0     2060
526413       480       3  974631272.0     2060
526414      1573       2  974631272.0     2060
526415      1676       3  974631272.0     2060
526434      2763       1  974631301.0     2060
526488       608       5  974631402.0     2060
526507       593       3  974631446.0     2060
526508       296       4  974631446.0     2060
526519      1358       3  974631471.0     2060
...          ...     ...          ...      ...
537887      2335       3  974655912.0     2060
537888       415       3  974655912.0     2060
537906      1100       3  974655941.0     2060
537920      1644       3  974655974.0     2060
537942      2266       2  974656010.0     2060
537991       765       3  974656077.0     2060
537992      1831       2  974656078.0     2060
538018       420       2  974656109.0     2060
538019       234       4  974656109.0     2060
538101       172       2  974656227.0     2060
538121       291       3  974656268.0     2060
538122        65       3  974656268.0     2060
538153       193       3  974656300.0     2060
538154      3835       3  974656300.0     2060
538155      1474       2  974656300.0     2060
538185       438       3  974656356.0     2060
538195      2953       2  974656380.0     2060
538286       747       2  974656501.0     2060
538327      2606       3  974656584.0     2060
538376      1556       1  974656650.0     2060
538441       255       2  974656725.0     2060
538443       427       2  974656725.0     2060
538469      3802       1  974656758.0     2060
538470      2412       1  974656758.0     2060
538550      2007       3  974656916.0     2060
538794      3268       1  974657325.0     2060
538855      2568       2  974657436.0     2060
687646      3160       2  975341492.0     2060
687654      3114       5  975341550.0     2060
687657      3793       4  975341575.0     2060

[212 rows x 4 columns]

In [13]:
drf.movie_id.value_counts()

2858    2619
260     2258
1196    2244
1210    2204
589     2069
480     2025
2028    2003
1270    1993
608     1942
2571    1937
1580    1931
593     1917
1198    1900
110     1877
2762    1839
2396    1819
1197    1765
527     1754
1617    1734
1097    1725
2628    1722
2997    1706
858     1688
1265    1680
318     1670
2716    1639
356     1631
296     1626
1240    1586
1       1559
        ... 
3295       1
3601       1
2510       1
3290       1
730        1
1470       1
758        1
1558       1
1118       1
2845       1
796        1
3353       1
601        1
792        1
790        1
1115       1
789        1
3164       1
1630       1
3607       1
607        1
3337       1
98         1
2484       1
2308       1
2584       1
3842       1
3687       1
2811       1
3890       1
Name: movie_id, Length: 3642, dtype: int64

It looks like we have ratings for 3642 movies.

In [14]:
drf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 719949 entries, 0 to 719948
Data columns (total 4 columns):
movie_id     719949 non-null int64
rating       719949 non-null int64
timestamp    719949 non-null float64
user_id      719949 non-null int64
dtypes: float64(1), int64(3)
memory usage: 22.0 MB


In [15]:
drf.user_id.value_counts()

1680    1671
889     1384
4169    1300
4277    1255
3618    1198
1941    1183
1150    1175
5795    1159
4344    1145
4510    1114
4227    1091
3841    1083
4508    1054
3808    1048
5367    1006
3539     971
2063     963
5831     947
3224     923
4725     914
5643     913
3032     910
4064     898
2909     894
5954     887
1285     886
4447     878
1181     875
1980     871
3778     866
        ... 
1447       5
5484       5
5998       5
1547       5
4610       5
2775       5
1463       5
5291       5
918        5
3991       5
4418       5
4744       5
750        5
2057       5
4467       5
1252       5
3152       4
1779       4
4528       4
4463       4
827        4
2991       4
988        4
1256       3
3288       3
1551       3
4273       3
3038       2
3326       1
3459       1
Name: user_id, Length: 5400, dtype: int64

It looks like we 5400 unique users.

### Requests Dataframe. This is the Dataframe of movies users have not seen, which we will populate with recommendations.

In [16]:
#Notice this is a local file. There are computer clusters don't have local files so be careful.
rq = spark.read.json('data/requests.json')

In [17]:
#look at schema
rq

DataFrame[movie_id: bigint, rating: double, timestamp: double, user_id: bigint]

In [18]:
rq.limit(5).show()

+--------+------+------------+-------+
|movie_id|rating|   timestamp|user_id|
+--------+------+------------+-------+
|    2019|   NaN|9.56678777E8|   6040|
|     759|   NaN|9.56679248E8|   6040|
|    2858|   NaN|9.56679275E8|   6040|
|     246|   NaN|9.56679413E8|   6040|
|    1617|   NaN|9.56679473E8|   6040|
+--------+------+------------+-------+



In [19]:
rq.limit(5).toPandas()

movie_id  rating    timestamp  user_id
0      2019     NaN  956678777.0     6040
1       759     NaN  956679248.0     6040
2      2858     NaN  956679275.0     6040
3       246     NaN  956679413.0     6040
4      1617     NaN  956679473.0     6040

In [20]:
#Instead of loading this file in over again every time we want to look at it, we'll use
rq.persist()

DataFrame[movie_id: bigint, rating: double, timestamp: double, user_id: bigint]

In [21]:
rq.head()

Row(movie_id=2019, rating=nan, timestamp=956678777.0, user_id=6040)

In [22]:
drq = rq.toPandas()

In [23]:
drq.head()

movie_id  rating    timestamp  user_id
0      2019     NaN  956678777.0     6040
1       759     NaN  956679248.0     6040
2      2858     NaN  956679275.0     6040
3       246     NaN  956679413.0     6040
4      1617     NaN  956679473.0     6040

In [45]:
drq.loc[drq.user_id == 2060]

movie_id  rating     timestamp  user_id
58764       2421     NaN  9.746303e+08     2060
58810       1610     NaN  9.746312e+08     2060
58851       3755     NaN  9.746321e+08     2060
58899       2762     NaN  9.746328e+08     2060
58914       2134     NaN  9.746332e+08     2060
58928        780     NaN  9.746335e+08     2060
58930       2628     NaN  9.746336e+08     2060
59900       1617     NaN  9.746512e+08     2060
59941        543     NaN  9.746524e+08     2060
59942       2572     NaN  9.746524e+08     2060
59946        785     NaN  9.746526e+08     2060
59949       3686     NaN  9.746528e+08     2060
59953       3107     NaN  9.746528e+08     2060
59955       3869     NaN  9.746528e+08     2060
59958       2002     NaN  9.746529e+08     2060
172129      3578     NaN  9.777649e+08     2060
172130      3861     NaN  9.777649e+08     2060
172131      3510     NaN  9.777650e+08     2060
172132      1302     NaN  9.777651e+08     2060
172133      2012     NaN  9.777652e+08     2060
172134      2011     NaN  9.777652e+08     2060
172332      3173     NaN  9.777810e+08     2060
176153      3753     NaN  9.779500e+08     2060
178860      3617     NaN  9.780441e+08     2060
180901      3556     NaN  9.781178e+08     2060
180904      3785     NaN  9.781178e+08     2060
192022      3863     NaN  9.790607e+08     2060
192023      3826     NaN  9.790607e+08     2060
197840      3752     NaN  9.800211e+08     2060
207469      1073     NaN  9.821488e+08     2060
208484      2959     NaN  9.824833e+08     2060
210762      3879     NaN  9.833032e+08     2060
210763      3882     NaN  9.833032e+08     2060
220609      3317     NaN  9.876613e+08     2060
220610      3897     NaN  9.876614e+08     2060
225285      2502     NaN  9.901080e+08     2060
225286      3827     NaN  9.901080e+08     2060
226926      2769     NaN  9.910748e+08     2060
242199        50     NaN  9.998441e+08     2060
242660      1483     NaN  1.000475e+09     2060
242668      3911     NaN  1.000538e+09     2060
242748      1060     NaN  1.000667e+09     2060
255138      1235     NaN  1.011723e+09     2060
255139      2858     NaN  1.011723e+09     2060
258737       858     NaN  1.015014e+09     2060
259039      1221     NaN  1.015183e+09     2060
273228      3007     NaN  1.034743e+09     2060
273229       562     NaN  1.034743e+09     2060
273300      1954     NaN  1.035062e+09     2060

In [25]:
drq.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280260 entries, 0 to 280259
Data columns (total 4 columns):
movie_id     280260 non-null int64
rating       0 non-null float64
timestamp    280260 non-null float64
user_id      280260 non-null int64
dtypes: float64(2), int64(2)
memory usage: 8.6 MB


In [26]:
drq.movie_id.value_counts()

2858    809
3578    759
1196    746
260     733
1210    679
593     661
2571    653
2028    650
480     647
2762    620
1198    614
1580    607
1265    598
1270    590
589     580
3751    577
608     571
110     566
356     563
318     557
1617    554
1197    553
2396    550
527     550
296     545
1097    544
2716    542
3793    540
858     535
2997    535
       ... 
1324      1
865       1
775       1
744       1
3542      1
981       1
980       1
1908      1
1039      1
33        1
3209      1
1160      1
2592      1
712       1
3460      1
1154      1
3533      1
717       1
2850      1
2172      1
1316      1
634       1
1145      1
889       1
120       1
3280      1
2343      1
887       1
2258      1
402       1
Name: movie_id, Length: 3564, dtype: int64

It looks like we have 3540 unique movies.

In [27]:
drq.user_id.value_counts()

424     1226
549     1152
1088    1074
1448    1025
524     1016
4169    1014
1447     985
550      967
678      945
352      870
531      867
482      825
195      822
216      794
302      788
1733     777
329      764
245      764
1912     746
509      723
319      721
5394     697
53       684
543      671
202      670
1181     646
1425     644
411      635
1266     633
148      624
        ... 
1126       1
3897       1
1089       1
2081       1
3282       1
4264       1
1801       1
3541       1
2260       1
2516       1
1237       1
4941       1
4672       1
3446       1
835        1
1911       1
2568       1
799        1
5589       1
5206       1
2007       1
5898       1
1703       1
4570       1
867        1
2470       1
5796       1
986        1
5336       1
4652       1
Name: user_id, Length: 5970, dtype: int64

It looks like we have 5970 users to whom we want to recommend movies.

## Users Dataframe

In [41]:
usdf_headers = ['user_id', 'gender', 'age', 'occupation', 'zip_code']

In [48]:
usdf = pd.read_csv('data/users.dat', sep='::', names=usdf_headers)

/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


In [49]:
usdf.head()

user_id gender  age  occupation zip_code
0        1      F    1          10    48067
1        2      M   56          16    70072
2        3      M   25          15    55117
3        4      M   45           7    02460
4        5      M   25          20    55455

In [88]:
# zip codes aren't doing much for us so we'll drop them
usdf = usdf.drop('zip_code', axis=1)

In [89]:
usdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6040 entries, 0 to 6039
Data columns (total 4 columns):
user_id       6040 non-null int64
gender        6040 non-null object
age           6040 non-null int64
occupation    6040 non-null int64
dtypes: int64(3), object(1)
memory usage: 188.8+ KB


## Movies DataFrame

In [30]:
mvdf_headers = ['title', 'genres']

In [31]:
mvdf = pd.read_csv('data/movies.dat', sep='::', names=mvdf_headers, index_col=0)

/anaconda3/envs/learn-env/lib/python3.6/site-packages/ipykernel_launcher.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  """Entry point for launching an IPython kernel.


In [32]:
mvdf.head()

title                        genres
1                    Toy Story (1995)   Animation|Children's|Comedy
2                      Jumanji (1995)  Adventure|Children's|Fantasy
3             Grumpier Old Men (1995)                Comedy|Romance
4            Waiting to Exhale (1995)                  Comedy|Drama
5  Father of the Bride Part II (1995)                        Comedy

In [33]:
mvdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3883 entries, 1 to 3952
Data columns (total 2 columns):
title     3883 non-null object
genres    3883 non-null object
dtypes: object(2)
memory usage: 91.0+ KB


In [85]:
mvdf['genres'] = mvdf['genres'].apply(lambda x: x.split('|'))
mvdf.head()

title                            genres
1                    Toy Story (1995)   [Animation, Children's, Comedy]
2                      Jumanji (1995)  [Adventure, Children's, Fantasy]
3             Grumpier Old Men (1995)                 [Comedy, Romance]
4            Waiting to Exhale (1995)                   [Comedy, Drama]
5  Father of the Bride Part II (1995)                          [Comedy]

In [87]:
all_genres = set([item for movie in mvdf['genres'] for item in movie])

## Metadata Dataframe

In [35]:
metadf = pd.read_csv('data/movies_metadata.csv')

/anaconda3/envs/learn-env/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [36]:
metadf.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview    ...     release_date  \
0  Led by Woody, Andy's toys live happily in his ...    ...       1995-10-30   
1  When siblings Judy and Peter discover an encha...    ...       1995-12-15   
2  A family wedding reignites the ancient feud be...    ...       1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...    ...       1995-12-22   
4  Just when George Banks has recovered from his ...    ...       1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [37]:
#df['Capital'] = df['Country'].map(country_capital)
metadf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
adult                    45466 non-null object
belongs_to_collection    4494 non-null object
budget                   45466 non-null object
genres                   45466 non-null object
homepage                 7782 non-null object
id                       45466 non-null object
imdb_id                  45449 non-null object
original_language        45455 non-null object
original_title           45466 non-null object
overview                 44512 non-null object
popularity               45461 non-null object
poster_path              45080 non-null object
production_companies     45463 non-null object
production_countries     45463 non-null object
release_date             45379 non-null object
revenue                  45460 non-null float64
runtime                  45203 non-null float64
spoken_languages         45460 non-null object
status                   45379 non-null objec

In [38]:
print(metadf.genres[1])
print(type(metadf.genres[1]))

[{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10751, 'name': 'Family'}]
<class 'str'>


## TrainTestSplit and OneHotEncoding

In [90]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [ ]:
Users_train, Users_test, rating_train, rating_test = train_test_split(usdf, )

In [91]:
def ohe_columns(series, name):
    ohe = OneHotEncoder(categories='auto')
    ohe.fit(series)
    cols = ohe.get_feature_names(name)
    ohe = ohe.transform(series)
    final_df = pd.DataFrame(ohe.toarray(), columns=cols)
    return final_df

In [93]:
# OHE the user cols
my_cols = ['gender', 'age', 'occupation']

ohe_multi = OneHotEncoder(categories='auto')
ohe_multi.fit(usdf[my_cols])
ohe_mat = ohe_multi.transform(usdf[my_cols])

# Then KMeans cluster

k_clusters.fit(ohe_mat)

preds = k_clusters.predict(ohe_mat)

NameError: name 'KMeans' is not defined

## Clustering Demographics

In [94]:
from sklearn.cluster import KMeans

In [95]:
k_clusters = KMeans(n_clusters=8, random_state=42)

In [96]:
# cluster user demographics
k_clusters.fit(ohe_mat)
demo_preds = k_clusters.predict(ohe_mat)

In [98]:
demo_preds.shape

(6040,)

In [103]:
dcdf = pd.concat(ohe_multi.transform(usdf[my_cols]), demo_preds)

TypeError: cannot concatenate object of type "<class 'scipy.sparse.csr.csr_matrix'>"; only pd.Series, pd.DataFrame, and pd.Panel (deprecated) objs are valid

## Clustering

In [ ]:
#create dictionary of user ids and their clusters
cluster_dict = {}
for k, v in zip(user_df['id'].tolist(), user_df['cluster'].tolist()):
    cluster_dict[k] = v

### Trying to import databases differently

In [39]:
import json

In [40]:
# json.loads(metadf.genres[0])

JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 3 (char 2)

In [51]:
demo_ratings = pd.merge(usdf, drf, how='left', on='user_id')
demo_ratings.head()

user_id gender  age  occupation zip_code  movie_id  rating  timestamp
0        1      F    1          10    48067       NaN     NaN        NaN
1        2      M   56          16    70072       NaN     NaN        NaN
2        3      M   25          15    55117       NaN     NaN        NaN
3        4      M   45           7    02460       NaN     NaN        NaN
4        5      M   25          20    55455       NaN     NaN        NaN

## Try using SQL language to create database grouped by user_id

In [62]:
import pyspark
import pyspark.sql.functions as F

In [83]:
def show(jf, n=5):
    '''load portion of json dataframe into a pandas dataframe'''
    return jf.limit(n).toPandas()

In [70]:
# 
ratings_jf = rf[['user_id', 'movie_id', 'rating']]

In [84]:
show(ratings_jf)

user_id  movie_id  rating
0     6040       858       4
1     6040      2384       4
2     6040       593       5
3     6040      1961       4
4     6040      1419       3

In [72]:
ratings_jf.createOrReplaceTempView('rf')

In [75]:
user_rating_query = '''
SELECT user_id
, rating
, movie_id
FROM rf
'''

In [78]:
ratings_table = spark.sql(user_rating_query)

In [79]:
show(ratings_table)

user_id  rating  movie_id
0     6040       4       858
1     6040       4      2384
2     6040       5       593
3     6040       4      1961
4     6040       3      1419

In [80]:
show(ratings_table)

user_id  rating  movie_id
0     6040       4       858
1     6040       4      2384
2     6040       5       593
3     6040       4      1961
4     6040       3      1419